plugins/scaffolder-node/src/filters/createTemplateFilter.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { JsonValue } from '@backstage/types';
+import { Schema } from 'jsonschema';
+import { mapValues, pickBy } from 'lodash';
+import { z } from 'zod';
+import zodToJsonSchema from 'zod-to-json-schema';
+import { CreatedTemplateFilter, TemplateFilterExample } from './types';
+
+const jsonSchema = (schemaSpec: Schema | z.ZodType) => {
+  return schemaSpec instanceof z.ZodType
+    ? (zodToJsonSchema(schemaSpec) as Schema)
+    : schemaSpec;
+};
+
+/** @public */
+export type TemplateFilterOptions<
+  TInputSchema extends Schema | z.ZodType = {},
```

Comment:
Let's drop the `jsonschema#Schema` type from this, and just use `zod` instead.

I'm actually thinking that it might be worth us following a similar pattern to what we've done for the frontend system where it's actually of this type:

https://github.com/backstage/backstage/blob/8f85ab3b04e3311171ca281d722c5177a5bba576/packages/frontend-plugin-api/src/wiring/createExtensionBlueprint.ts#L55

So it means that it's defined like so:

```ts
schema: {
  input: (z) => z.object(...)
}
```

It's not exactly the same, but I think its a little nicer this way and we will be moving the `createScaffolderFieldExtension` to follow a similar pattern I think

Reply:
I don't hate this, but 1. wouldn't we want to restrict the keys, and 2. does this mean we'd be putting the burden on the caller to supply `arguments` as a tuple type (could automatically interpret a non-tuple argument as being unary)?